pkg/apis/provisioning/v1alpha5/constraints.go

```diff
@@ -47,26 +49,62 @@ type Provider = runtime.RawExtension
 
 // ValidatePod returns an error if the pod's requirements are not met by the constraints
 func (c *Constraints) ValidatePod(pod *v1.Pod) error {
+	p := pod.DeepCopy()
+	// The soft preference may conflict with the requirements.
+	// Remove soft constraints/ preferences
+	if p.Spec.Affinity != nil && p.Spec.Affinity.NodeAffinity != nil {
+		p.Spec.Affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution = nil
+	}
 	// Tolerate Taints
-	if err := c.Taints.Tolerates(pod); err != nil {
+	if err := c.Taints.Tolerates(p); err != nil {
 		return err
 	}
+	requirements := NewPodRequirements(p)
+	// Test if labels are allowed
+	for key := range requirements.Keys() {
+		if err := IsRestrictedLabel(key); err != nil && !PodLabelExceptions.Has(key) {
```

Comment:
Do we even need to validate this? These labels are combined with the provisioner. If the pod requests labels that the provisioner doesn't support, scheduling can't happen.

Reply:
This validation function is also used to validate provisioner. Now we allow the provisioner to use arbitrary labels. We need to make sure that reserved labels like those of k8s.io/XXX are not used.

Reply:
We're using ValidatePod to validate the provisioner?

Reply:
I mis-spoken. Given that all keys will be defined in the provisioner, we are fine to not check during pod validation.